examples/tqc/utils.py

```diff
@@ -0,0 +1,410 @@
+    """
+    pairwise_delta = samples[:, None, None, :] - quantiles[:, :, :, None]  # batch x nets x quantiles x samples
+    abs_pairwise_delta = torch.abs(pairwise_delta)
+    huber_loss = torch.where(abs_pairwise_delta > 1,
+                             abs_pairwise_delta - 0.5,
+                             pairwise_delta ** 2 * 0.5)
+    n_quantiles = quantiles.shape[2]
+    tau = torch.arange(n_quantiles, device=quantiles.device).float() / n_quantiles + 1 / 2 / n_quantiles
+    loss = (torch.abs(tau[None, None, :, None] - (pairwise_delta < 0).float()) * huber_loss).mean()
+    return loss
+
+
+# ====================================================================
+# TQC Loss
+# --------
+
+class TQCLoss(LossModule):
+    def __init__(
+            self,
+            actor_network,
+            qvalue_network,
+            gamma,
+            top_quantiles_to_drop,
+            alpha_init,
+            device
+    ):
+        super(type(self), self).__init__()
+        super().__init__()
+
+        self.convert_to_functional(
+            actor_network,
+            "actor",
+            create_target_params=False,
+            funs_to_decorate=["forward", "get_dist"],
+        )
+
+        self.convert_to_functional(
+            qvalue_network,
+            "critic",
+            create_target_params=True  # Create a target critic network
+        )
+
+        self.device = device
+        self.log_alpha = torch.tensor([np.log(alpha_init)], requires_grad=True, device=self.device)
+        self.gamma = gamma
+        self.top_quantiles_to_drop = top_quantiles_to_drop
+
+        # Compute target entropy
+        action_spec = getattr(self.actor, "spec", None)
+        if action_spec is None:
+            print("Could not deduce action spec from actor network.")
+        if not isinstance(action_spec, CompositeSpec):
+            action_spec = CompositeSpec({"action": action_spec})
+        action_container_len = len(action_spec.shape)
+        self.target_entropy = -float(action_spec["action"].shape[action_container_len:].numel())
+
+    def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
+        td_next = tensordict["next"]
+        reward = td_next["reward"]
+        not_done = tensordict["done"].logical_not()
+
+        alpha = torch.exp(self.log_alpha)
+
+        # Q-loss
+        with torch.no_grad():
+            # get policy action
+            td_next = self.actor(td_next, params=self.actor_params)
+            td_next = self.critic(td_next, params=self.target_critic_params)
+
+            next_log_pi = td_next["sample_log_prob"]
+            next_log_pi = torch.unsqueeze(next_log_pi, dim=-1)
+
+            # compute and cut quantiles at the next state
+            next_z = td_next["state_action_value"]
+            sorted_z, _ = torch.sort(next_z.reshape(next_z.shape[0], -1))
+            sorted_z_part = sorted_z[:, :-self.top_quantiles_to_drop]
```

Comment:
Probably worth considering what happens if the batch size is greater or smaller than 1 (eg `[..., :-top_quantiles_to_drop]`